data/trie/trieStorageManager.go

```diff
@@ -275,32 +275,44 @@ func (tsm *trieStorageManager) snapshot(msh marshal.Marshalizer, hsh hashing.Has
 		tr, err := newSnapshotTrie(tsm.db, msh, hsh, snapshot.rootHash)
 		if err != nil {
 			tsm.storageOperationMutex.Unlock()
+			isSnapshotsBufferEmpty, keys = tsm.isSnapshotsBufferEmpty()
+
 			log.Error("trie storage manager: newSnapshotTrie", "error", err.Error())
-			return
+			break
 		}
 		db := tsm.getSnapshotDb(snapshot.newDb)
 
 		tsm.storageOperationMutex.Unlock()
 
 		err = tr.root.commit(true, 0, tsm.db, db)
 		if err != nil {
+			isSnapshotsBufferEmpty, keys = tsm.isSnapshotsBufferEmpty()
+
 			log.Error("trie storage manager: commit", "error", err.Error())
-			return
+			break
 		}
 
-		tsm.storageOperationMutex.Lock()
-		tsm.snapshotsBuffer.removeFirst()
-		isSnapshotsBufferEmpty = tsm.snapshotsBuffer.len() == 0
-		if isSnapshotsBufferEmpty {
-			keys = tsm.pruningBuffer
-			tsm.pruningBuffer = make([][]byte, 0)
-		}
-		tsm.storageOperationMutex.Unlock()
+		isSnapshotsBufferEmpty, keys = tsm.isSnapshotsBufferEmpty()
 	}
 
 	tsm.removeKeysFromDb(keys)
 }
 
+func (tsm *trieStorageManager) isSnapshotsBufferEmpty() (bool, [][]byte) {
+	tsm.storageOperationMutex.Lock()
+	defer tsm.storageOperationMutex.Unlock()
```

Comment:
You can use tsm.storageOperationMutex.Unlock() in line 312 instead defer

Reply:
Done.